for i in {1..7}
do
    echo -e '#!/bin/bash\nstart=`date +%s` \n## This script is used to deduplicate bam files and clipped overlapping read pairs for paired end data. It can process both paired end and single end data.\nBAMLIST=/workdir/hz269/DelBay20/sample_lists/bam_list_'$i'.txt # Path to a list of merged, deduplicated, and overlap clipped bam files. Full paths should be included.\nSAMPLETABLE=$BASEDIR/sample_lists/fastq_table.txt # Path to a sample table where the 1st column is the prefix of the raw fastq files. The 4th column is the sample ID, the 2nd column is the lane number, and the 3rd column is sequence ID. The combination of these three columns have to be unique. The 6th column should be data type, which is either pe or se.\nPICARD=/programs/picard-tools-2.19.2/picard.jar # Path to picard tools\nBAMUTIL=/programs/bamUtil/bam # Path to bamUtil\ncd bam\n\n## Loop over each sample\nfor SAMPLEBAM in `cat $BAMLIST`; do\n\n\t## Extract the file name prefix for this sample\n\tSAMPLESEQID=`echo $SAMPLEBAM | sed '\''s/_bt2_.*//'\'' | sed -e '\''s#.*/bam/\(\)#\1#'\''`\n\tSAMPLEPREFIX=`echo ${SAMPLEBAM%.bam}`\n\n\t## Remove duplicates and print dupstat file\n\t# We used to be able to just specify picard.jar on the CBSU server, but now we need to specify the path and version\n\tjava -Xmx60g -jar $PICARD MarkDuplicates I=$SAMPLEBAM O=$SAMPLEPREFIX'\''_dedup.bam'\'' M=$SAMPLEPREFIX'\''_dupstat.txt'\'' VALIDATION_STRINGENCY=SILENT REMOVE_DUPLICATES=true\n\n\t## Clip overlapping paired end reads (only necessary for paired end data)\n\t$BAMUTIL clipOverlap --in $SAMPLEPREFIX'\''_dedup.bam'\'' --out $SAMPLEPREFIX'\''_dedup_overlapclipped.bam'\'' --stats\n\ndone\n\nend=`date +%s` ## date at end\nruntime=$((end-start))\nhours=$((runtime / 3600))\nminutes=$(( (runtime % 3600) / 60 ))\nseconds=$(( (runtime % 3600) % 60 ))\necho "Runtime: $hours:$minutes:$seconds (hh:mm:ss)"' >> formal/'8_dup_clip_'$i'.sh'
done
