for i in {1..1}
do
    echo -e '#!/bin/bash\nstart=`date +%s`\nBAMLIST=/workdir/hz269/DelBay_test/sample_lists/bam_list_dedup_overlapclipped.list # Path to a list of merged, deduplicated, and overlap clipped bam files. Full paths should be included.\nBASEDIR=/workdir/hz269/DelBay_test\nREFERENCE=$BASEDIR/reference/CV30_masked.fasta # Path to reference fasta file and file name\nSAMTOOLS=/programs/samtools-1.11/bin/samtools # Path to samtools\nGATK=/programs/GenomeAnalysisTK-3.7/GenomeAnalysisTK.jar # Path to GATK\n\ncd bam\n\n## Loop over each sample\nfor SAMPLEBAM in `cat $BAMLIST`; do\n\nif [ -e $SAMPLEBAM'\''.bai'\'' ]; then\n\techo "the file already exists"\nelse\n\t## Index bam files\n\t$SAMTOOLS index $SAMPLEBAM\nfi\ndone\n\n## Realign around in-dels\n# This is done across all samples at once\n\n## Use an older version of Java\nexport JAVA_HOME=/usr/local/jdk1.8.0_121\nexport PATH=$JAVA_HOME/bin:$PATH\n\n## Create list of potential in-dels\nif [ ! -f $BASEDIR'\''/bam/all_samples_for_indel_realigner.intervals'\'' ]; then\n\t\tjava -Xmx120g -jar $GATK \\\n\t\t\t-T RealignerTargetCreator \\\n\t\t\t-R $REFERENCE \\\n\t\t\t-I $BAMLIST \\\n\t\t\t-o $BASEDIR'\''/bam/all_samples_for_indel_realigner.intervals'\'' \\\n\t\t\t-drf BadMate\nfi\n\n## Run the indel realigner tool\njava -Xmx120g -jar $GATK \\\n\t-T IndelRealigner \\\n\t-R $REFERENCE \\\n\t-I $BAMLIST \\\n\t-targetIntervals $BASEDIR'\''/bam/all_samples_for_indel_realigner.intervals'\'' \\\n\t--consensusDeterminationModel USE_READS  \\\n\t--nWayOut _realigned.bam\n\nend=`date +%s` ## date at end\nruntime=$((end-start))\nhours=$((runtime / 3600))\nminutes=$(( (runtime % 3600) / 60 ))\nseconds=$(( (runtime % 3600) % 60 ))\necho "Runtime: $hours:$minutes:$seconds (hh:mm:ss)"' >> formal/'9_realign_'$i'.sh'
done
