for i in {1..19}
do
    echo -e '#!/bin/bash\nstart=`date +%s`\n#!/bin/bash\n\n## This script is used to count number of bases in raw, adapter clipped, and quality filtered fastq files. The result of this script will be stored in a nohup file.\nBASEDIR=/workdir/hz269/DelBay20/\nSAMPLELIST=$BASEDIR/sample_lists/fastq_list_'$i'.txt\nSAMPLETABLE=$BASEDIR/sample_lists/fastq_table.txt\nRAWFASTQDIR=/workdir/hz269/DelBay20/raw_fastq/\nQUALFILTERED=true # Whether the sample has gone through quality filtering. true or false\n\n# Create headers for the output\nif $QUALFILTERED; then\n\tprintf '\''sample_seq_id\\traw_reads\\traw_bases\\tadapter_clipped_bases\\tqual_filtered_bases\\n'\''\nelse\n\tprintf '\''sample_seq_id\\traw_reads\\traw_bases\\tadapter_clipped_bases\\n'\''\nfi\n\n# Loop over each sample in the sample table\nfor SAMPLEFILE in `cat $SAMPLELIST`; do\n\tRAWFASTQFILES=$RAWFASTQDIR$SAMPLEFILE'\''*.gz'\''  # The input path and file prefix\n\n\t# Count the number of reads in raw fastq files. We only need to count the forward reads, since the reverse will contain exactly the same number of reads. fastq files contain 4 lines per read, so the number of total reads will be half of this line number.\n\tRAWREADS=`zcat $RAWFASTQFILES | wc -l`\n\n\t# Count the number of bases in raw fastq files. We only need to count the forward reads, since the reverse will contain exactly the same number of bases. The total number of reads will be twice this count.\n\tRAWBASES=`zcat $RAWFASTQFILES | awk '\''NR%4==2'\'' | tr -d "\\n" | wc -m`\n\n\t# Extract relevant values from a table of sample, sequencing, and lane ID (here in columns 4, 3, 2, respectively) for each sequenced library\n\tPOP_ID=`grep -P "${SAMPLEFILE}\\t" $SAMPLETABLE | cut -f 5`\n\tSAMPLE_ID=`grep -P "${SAMPLEFILE}\\t" $SAMPLETABLE | cut -f 4`\n\tSEQ_ID=`grep -P "${SAMPLEFILE}\\t" $SAMPLETABLE | cut -f 3`\n\tLANE_ID=`grep -P "${SAMPLEFILE}\\t" $SAMPLETABLE | cut -f 2`\n\tSAMPLE_SEQ_ID=$SAMPLE_ID'\''_'\''$POP_ID'\''_'\''$SEQ_ID'\''_'\''$LANE_ID\n\n\t# Find all adapter clipped fastq files corresponding to this sample and store them in the object ADAPTERFILES.\n\tADAPTERFILES=$BASEDIR'\''adapter_clipped/'\''$SAMPLE_SEQ_ID'\''*.gz'\''\n\n\t# Count all bases in adapter clipped files.\n\tADPTERCLIPBASES=`zcat $ADAPTERFILES | awk '\''NR%4==2'\'' | tr -d "\\n" | wc -m`\n\n\t# If reads are quality filtered, count quality filtered files.\n\tif $QUALFILTERED; then\n\n\t\t# Find all quality trimmed fastq files corresponding to this sample and store them in the object QUALFILES.\n\t\tQUALFILES=$BASEDIR'\''qual_filtered/'\''$SAMPLE_SEQ_ID'\''*.gz'\''\n\n\t\t# Count bases in quality trimmed files.\n\t\tQUALFILTPBASES=`zcat $QUALFILES | awk '\''NR%4==2'\'' | tr -d "\\n" | wc -m`\n\n\t\t# Write the counts in appropriate order.\n\t\tprintf "%s\\t%s\\t%s\\t%s\\t%s\\n" $SAMPLE_SEQ_ID $((RAWREADS/4)) $RAWBASES $ADPTERCLIPBASES $QUALFILTPBASES\n\n\t\t# When reads are not quality filtered, directly write the output\n\telse\n\n\t\t# Write the counts in appropriate order.\n\t\tprintf "%s\\t%s\\t%s\\t%s\s\\n" $SAMPLE_SEQ_ID $((RAWREADS/4)) $RAWBASES $ADPTERCLIPBASES\n\n\tfi\n\ndone\n\nend=`date +%s` ## date at end\nruntime=$((end-start))\nhours=$((runtime / 3600))\nminutes=$(( (runtime % 3600) / 60 ))\nseconds=$(( (runtime % 3600) % 60 ))\necho "Runtime: $hours:$minutes:$seconds (hh:mm:ss)"' >> formal/'count_fastq_'$i'.sh'
done
